llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -10137,8 +10140,17 @@ static const SCEV *SolveLinEquationWithOverflow(const APInt &A, const SCEV *B,
   //
   // B is divisible by D if and only if the multiplicity of prime factor 2 for B
   // is not less than multiplicity of this prime factor for D.
-  if (SE.getMinTrailingZeros(B) < Mult2)
-    return SE.getCouldNotCompute();
+  if (SE.getMinTrailingZeros(B) < Mult2) {
+    if (!Predicates)
+      return SE.getCouldNotCompute();
+    // Try to add a predicate ensuring B is a multiple of A.
+    const SCEV *URem = SE.getURemExpr(B, SE.getConstant(A));
```

Comment:
Why `B % A`, and not `B % (1 << Mult2)`?

I don't think this actually impacts any of the existing testcases; probably worth adding a few so we have coverage.

Reply:
Updated thanks. Test has been added in https://github.com/llvm/llvm-project/commit/28439a19c15f4273ea56342ac3e20a8918307561